test/cpp/tensorexpr/test_conv.cpp

```diff
@@ -0,0 +1,88 @@
+#include <gtest/gtest.h>
+#include "torch/csrc/jit/tensorexpr/ir_simplifier.h"
+#include "torch/csrc/jit/tensorexpr/loopnest.h"
+#include "torch/csrc/jit/tensorexpr/tensor.h"
+#include "torch/torch.h"
+
+namespace torch {
+namespace jit {
+
+namespace te = torch::jit::tensorexpr;
+namespace F = torch::nn::functional;
+
+void testConv2D() {
+  te::KernelScope kernel_scope;
+
+  // Input dimensions.
+  constexpr int N = 1;
+  constexpr int C = 3;
+  constexpr int H = 11;
+  constexpr int W = 11;
+
+  // Filter dimensions.
+  constexpr int K = 8;
+  constexpr int R = 3;
+  constexpr int S = 3;
+
+  // Output dims.
+  constexpr int OH = H - R + 1;
+  constexpr int OW = W - S + 1;
+
+  // Compute reference result.
+  at::Tensor input = torch::randn({N, C, H, W});
+  at::Tensor filter = torch::randn({K, C, R, S});
+  at::Tensor ref = F::conv2d(input, filter);
+
+  // Double check the output size is as expected.
+  ASSERT_EQ(ref.size(0), N);
+  ASSERT_EQ(ref.size(1), K);
+  ASSERT_EQ(ref.size(2), OH);
+  ASSERT_EQ(ref.size(3), OW);
+
+  te::Buffer inputB(te::BufHandle("input", {N, C, H, W}, te::kFloat));
+  te::Buffer filterB(te::BufHandle("filter", {K, C, R, S}, te::kFloat));
+
+  te::Tensor* conv = te::Reduce(
+      "conv",
+      {{N, "n"}, {K, "k"}, {OH, "oh"}, {OW, "ow"}},
+      te::Sum(),
+      // FIXME: We have to use a `std::vector` parameter here and then unpack
+      // it, because we don't have an overload allowing for an arbitrary number
+      // of ExprHandle/VarHandle parameters.
+      [&](const std::vector<te::VarHandle>& v) {
+        auto const& n = v[0];
+        auto const& k = v[1];
+        auto const& oh = v[2];
+        auto const& ow = v[3];
+        auto const& c = v[4];
+        auto const& r = v[5];
+        auto const& s = v[6];
+        // FIXME: We have to use `call` and construct a `std::vector` here
+        // because the `operator()` overload is only specialized for a small
+        // number of arguments.
+        return inputB.call(
+                   std::vector<te::ExprHandle>({n, c, oh + r, ow + s})) *
```

Comment:
We need the handling of the paddings here. I am a bit surprised that this didn't OOB....

Reply:
Hm really?  This compute expression is an unpadded conv (or at least it's supposed to be).  Since `OH=H-R+1` the max extent of `oh + r` would be. (\me does algebra) `(H-R+1)-1 + (R-1)` or `H-1`, which is in bounds.  Right?